src/riak_core_coverage_plan.erl

```diff
@@ -110,120 +293,433 @@ create_plan(VNodeSelector, NVal, PVC, ReqId, Service) ->
+    %% ID by one to find the relevant keyspace index for traditional
+    %% coverage plan filters
+    id_to_index(add_offset(PartitionID, 1, RingSize), RingSize);
+convert({partition_id, PartitionID, RingSize}, partition_index) ->
+    id_to_index(PartitionID, RingSize);
+convert({vnode_id, VNodeID, RingSize}, vnode_index) ->
+    id_to_index(VNodeID, RingSize);
+convert({SubpID, Bits}, subpartition_index) ->
+    SubpID bsl Bits.
+
+%% @private
+partition_to_preflist(Partition, NVal, Offset, UpNodes, CHBin) ->
+    DocIdx = convert(Partition, partition_index),
+    docidx_to_preflist(DocIdx, NVal, Offset, UpNodes, CHBin).
+
+%% @private
+docidx_to_preflist(DocIdx, NVal, Offset, UpNodes, CHBin) ->
+    %% `chashbin' would be fine with a straight integer, but
+    %% `riak_core_apl' has a -spec that mandates binary(), so we'll
+    %% play its game
+    OrigPreflist =
+        lists:map(fun({{Idx, Node}, primary}) -> {Idx, Node} end,
+                  riak_core_apl:get_primary_apl_chbin(
+                    <<DocIdx:160/integer>>, NVal, CHBin, UpNodes)),
+    rotate_list(OrigPreflist, length(OrigPreflist), Offset).
+
+%% @private
+rotate_list(List, Len, Offset) when Offset >= Len ->
+    List;
+rotate_list(List, _Len, Offset) ->
+    {Head, Tail} = lists:split(Offset, List),
+    Tail ++ Head.
+
+%% @private
+index_to_id(Index, RingSize) ->
+    RingIndexInc = chash:ring_increment(RingSize),
+    Index div RingIndexInc.
+
+%% @private
+id_to_index(Id, RingSize) when Id < RingSize->
+    RingIndexInc = chash:ring_increment(RingSize),
+    Id * RingIndexInc.
+
+%% @private
+find_vnode_partitions(Index, N, RingSize) ->
+    n_keyspaces({vnode_id, index_to_id(Index, RingSize), RingSize}, N).
+
+%% @private
+partitions_by_index_or_filter(Idx, NVal, [], RingSize) ->
+    find_vnode_partitions(Idx, NVal, RingSize);
+partitions_by_index_or_filter(_Idx, _NVal, Filters, RingSize) ->
+    %% Filters for traditional coverage plans are offset by 1, so when
+    %% converting to partition indexes or IDs, we have to subtract 1
+    lists:map(fun(P) ->
+                      {partition_id,
+                       add_offset(index_to_id(P, RingSize),
+                                  -1, RingSize),
+                       RingSize} end,
+              Filters).
+
+
+
+%% @private
+%% Must be able to comply with PVC if the target is 'all'
+check_pvc(List, _PVC, allup) ->
+    List;
+check_pvc(List, PVC, all) ->
+    check_pvc2(List, length(List), PVC).
+
+%% @private
+check_pvc2(List, Len, PVC) when Len >= PVC ->
+    List;
+check_pvc2(_List, _Len, _PVC) ->
+    [].
+
+%% @private
+create_subpartition_plan(VNodeTarget, NVal, Count, PVC, ReqId, Service, CHBin, AvailNodeFun) ->
+    MaskBSL = data_bits(Count),
+
+    %% Calculate an offset based on the request id to offer the
+    %% possibility of different sets of VNodes being used even when
+    %% all nodes are available.
+    Offset = ReqId rem NVal,
```

Comment:
Should this use a better source of randomness? Not maybe cryptographic strength randomness, but the request term is going to be fairly monotonic and not very random.